sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbStateStoreProvider.scala

```diff
@@ -0,0 +1,642 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming.state;
+
+import java.io._
+import java.util
+import java.util.Locale
+
+import scala.collection.JavaConverters._
+import scala.util.control.NonFatal
+
+import org.apache.commons.io.FileUtils
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+
+import org.apache.spark.{SparkConf, SparkEnv}
+import org.apache.spark.internal.Logging
+import org.apache.spark.io.FileUtility
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow
+import org.apache.spark.sql.execution.streaming.CheckpointFileManager
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.util.Utils
+
+private[sql] class RocksDbStateStoreProvider extends StateStoreProvider with Logging {
+
+  /* Internal fields and methods */
+  @volatile private var stateStoreId_ : StateStoreId = _
+  @volatile private var keySchema: StructType = _
+  @volatile private var valueSchema: StructType = _
+  @volatile private var storeConf: StateStoreConf = _
+  @volatile private var hadoopConf: Configuration = _
+  @volatile private var numberOfVersionsToRetain: Int = _
+  @volatile private var localDirectory: String = _
+
+  /*
+   * Additional configurations related to rocksDb. This will capture all configs in
+   * SQLConf that start with `spark.sql.streaming.stateStore.rocksDb`
+   */
+  @volatile private var rocksDbConf: Map[String, String] = Map.empty[String, String]
+
+  private lazy val baseDir: Path = stateStoreId.storeCheckpointLocation()
+  private lazy val fm = CheckpointFileManager.create(baseDir, hadoopConf)
+  private lazy val sparkConf = Option(SparkEnv.get).map(_.conf).getOrElse(new SparkConf)
+
+  private case class StoreFile(version: Long, path: Path, isSnapshot: Boolean)
+
+  import WALUtils._
+
+  /** Implementation of [[StateStore]] API which is backed by RocksDB and HDFS */
+  class RocksDbStateStore(val version: Long) extends StateStore with Logging {
+
+    /** Trait and classes representing the internal state of the store */
+    trait STATE
+
+    case object LOADED extends STATE
+
+    case object UPDATING extends STATE
+
+    case object COMMITTED extends STATE
+
+    case object ABORTED extends STATE
+
+    private val newVersion = version + 1
+    @volatile private var state: STATE = LOADED
+    private val finalDeltaFile: Path = deltaFile(baseDir, newVersion)
+    private lazy val deltaFileStream = fm.createAtomic(finalDeltaFile, overwriteIfPossible = true)
+    private lazy val compressedStream = compressStream(deltaFileStream, sparkConf)
+
+    override def id: StateStoreId = RocksDbStateStoreProvider.this.stateStoreId
+
+    var rocksDbWriteInstance: OptimisticTransactionDbInstance = null
```

Comment:
Option?

Reply:
I wanted to keep it similar to other RocksDbInstance objects. All operations on rocksDbWriteInstance would be better readable than it is if we use option.

Will make the change if you think we should be using option here.